### ISSUE TITLE
Compare Expression vs AddSubtract Performance

### ISSUE THREAD

**Issue by cxbrooks** [cxbrooks/test#36]
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#36
From: @cxbrooks
Reported version: 6.0-devel

**Comment by cxbrooks** [comment]
It would be interesting to compare Expression vs AddSubtract performance.

(From the July Rome AFRL Visit)

**Comment by cxbrooks** [comment]
The executive summary is that AddSubtract is about 20% faster than
Expression in my simple test case.

I created two models in $PTII/ptolemy/actor/lib/test/performance
that consist of


  Ramp ---->
             AddSubtract ---> Discard
  Ramp ---->

and

  Ramp ---->
             Expression ---> Discard
  Ramp ---->


Where the Ramps are producing integers.
When I run the models, I run for 100k iterations and
I throw away the data from the first run when computing the average time.

I get:
AddSubtractPerformance.xml: Average time of 2 through 5 runs: 696 ms.
ExpressionPerformance.xml: Average time of 2 through 5 runs: 866 ms.

This means that in this extreme case AddSubtract runs about 20% faster
than Expression when adding two integers.

Below is a sample run with Integers on my laptop using
MoMLSimpleApplication:

bash-3.1$ make
make runModel MODEL=AddSubtractPerformance.xml
make[1]: Entering directory `/cygdrive/c/cxh/pt
II/ptolemy/actor/lib/test/performance'
Run AddSubtractPerformance.xml 5 times
721 ms. Memory: 1984K Free: 1155K (58%)
701 ms. Memory: 1984K Free: 1155K (58%)
691 ms. Memory: 1984K Free: 1155K (58%)
691 ms. Memory: 1984K Free: 1155K (58%)
701 ms. Memory: 1984K Free: 1155K (58%)
AddSubtractPerformance.xml: Average time of 2 through 5 runs: 696 ms.
make[1]: Leaving directory `/cygdrive/c/cxh/pt
II/ptolemy/actor/lib/test/performance'
make runModel MODEL=ExpressionPerformance.xml
make[1]: Entering directory `/cygdrive/c/cxh/pt
II/ptolemy/actor/lib/test/performance'
Run ExpressionPerformance.xml 5 times
872 ms. Memory: 1984K Free: 1601K (81%)
881 ms. Memory: 1984K Free: 1601K (81%)
861 ms. Memory: 1984K Free: 1601K (81%)
861 ms. Memory: 1984K Free: 1601K (81%)
861 ms. Memory: 1984K Free: 1601K (81%)
ExpressionPerformance.xml: Average time of 2 through 5 runs: 866 ms.
make[1]: Leaving directory `/cygdrive/c/cxh/pt
II/ptolemy/actor/lib/test/performance'
bash-3.1$
Changing the step to doubles increases the time slightly but
not that much

           AddSubtract  Expression   AddSubtract/Expression
Integers     696        866          0.804
Integers     693.5 ms   856.25  ms   0.809
Integers     703.5      883.75       0.796

Doubles      711        883.75       0.805
Doubles      708        874          0.810
Doubles      701        878.5        0.798

This is basically a worst case situation, where we are attempting
to run the models flat out.

I ran the C code generator on these models and I get
times of 50ms for AddSubtract and 60ms for Expression, a speed
up on the order of 15x.

bash-3.1$ time ./AddSubtractPerformance.exe

real      0m0.050s
user      0m0.030s
sys       0m0.030s
bash-3.1$ time ./ExpressionPerformance.exe

real      0m0.060s
user      0m0.020s
sys       0m0.050s

Of course, the problem here is that the accuracy of the Windows
clock is pretty small.  To get really good numbers, we should increase
the number of iterations.

We could do more here, like profiling, but would be a separate bug.

_Christopher

  "There are lies, damn lies and statistics"
                                              - Mark Twain
